File: largestValueLevel.java
// Time Complexity : O(N) N = number of Nodes
// Space Complexity : O(H) H = number of nodes in level with maxi,u, elements
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/*
Do levelorder traversal and store the largestValues in arraylist
*/

// Your code here along with comments explaining your approach
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        if(root == null) return list;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        
        while(!q.isEmpty()) {
            int length = q.size();
            int maxLevelValue = Integer.MIN_VALUE;
            for(int i=0;i<length;i++) {
                TreeNode curr = q.poll();
                if(curr.val > maxLevelValue) maxLevelValue = curr.val;
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
                
            }
            list.add(maxLevelValue);
        }
        return list;
    }
}

File: Largest_value_level.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#Time:O(n) As we are iterating through the whole tree
#Space: O(n) we are using a queue to store all the elements
from collections import deque
class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        self.ans = []
        queue = deque()
        queue.append((root,0))
        while(queue):
            root,index = queue.popleft()
            if len(self.ans)==index+1:
                self.ans[index]=max(self.ans[index],root.val)
            else:
                self.ans.append(root.val)
            if root.left:
                queue.append((root.left,index+1))
            if root.right:
                queue.append((root.right,index+1))
        return self.ans
        